src/components/sharing/util.js

```diff
@@ -0,0 +1,431 @@
+    ];
+};
+
+const addDisplayPermissions = (userMap, resourceTotal) => {
+    return Object.keys(userMap).reduce(
+        (acc, userId) => {
+            const user = acc[userId];
+            const resources = getUserResourceList(user);
+
+            const permissions = resources.map(
+                (resource) => resource.permission
+            );
+            const displayPermission = getDisplayPermission(
+                permissions,
+                resourceTotal
+            );
+            return {
+                ...acc,
+                [userId]: {
+                    ...user,
+                    displayPermission,
+                    originalPermission: displayPermission,
+                },
+            };
+        },
+        { ...userMap }
+    );
+};
+
+/**
+ * This will create a mapping where each key is a user ID.  The value is
+ * the user's information from the user-info service with 3 things added:
+ * - originalPermission: contains the original overall permission for the user
+ *
+ * - displayPermission: contains the current overall permission for the user.
+ * If the user's permission gets updated in the sharing dialog, this value will
+ * update.
+ *
+ * - paths, apps, analyses, tools keys for storing the array of resources the
+ * user has or will have permissions on
+ *
+ * @param responses - permission-lister responses
+ * @param userInfos
+ * @param resourceTotal
+ * @returns {{}}
+ */
+export const getUserMap = (responses, userInfos, resourceTotal) => {
+    const userMap = responses.reduce((acc, resp) => {
+        const type = getResponseType(resp);
+        const { permListFn, userIdFn, fmtSharingFn } = getSharingFns(type);
+        return resp[type].reduce(
+            (shares, sharing) => {
+                const permissions = permListFn(sharing);
+
+                return permissions.reduce(
+                    (userMap, permission) => {
+                        const userId = userIdFn(permission);
+                        const currentUser = {
+                            ...(userMap[userId] || userInfos[userId]),
+                        };
+                        const permissionValue = permission.permission;
+
+                        // Create an obj with the resource details and the user's
+                        // current permission
+                        const resourceWithPermission = {
+                            ...fmtSharingFn(sharing),
+                            permission: permissionValue,
+                        };
+
+                        const updatedUser = {
+                            ...currentUser,
+                            [type]: [
+                                ...(currentUser[type] || []),
+                                resourceWithPermission,
+                            ],
+                        };
+                        return { ...userMap, [userId]: updatedUser };
+                    },
+                    { ...shares }
```

Comment:
I don't think this copy is necessary either.

Reply:
😄 I just love making copies I guess 🤷‍♀️